Library/Homebrew/dependencies_helpers.rb

```diff
@@ -49,6 +59,12 @@ def recursive_includes(klass, root_dependent, includes, ignores)
         keep ||= dep.test? && includes.include?("test?") && dependent == root_dependent
         keep ||= dep.build? && includes.include?("build?")
         klass.prune unless keep
+
+        next unless skip_recursive_build_dependents
+        next unless dep.build?
+        next if dependent == root_dependent && used_formulae.include?(dep.to_formula)
```

Comment:
I'm not clear on how this works. When would `used_formulae` contain a dependency? Is this when you specify multiple formulae?

My initial thought here was this would be `Dependency.keep_but_prune_recursive_deps`.

Reply:
I'll confess to still being confused about how this works. What I know is that this seems to produce what I'm looking for (with some caveats when we pass `--build-dependents-from-source` to `test-bot`).

> My initial thought here was this would be `Dependency.keep_but_prune_recursive_deps`.

I thought this was what I wanted too, but when I tried it the results didn't look right. But that might have been when I had the wrong `cache_key`. (I did do testing with no caching at all to be sure, but maybe I missed this variation.) I'll try it again to double check.

Reply:
> When would `used_formulae` contain a dependency?

`used_formulae` is an `Array` of `Formula`e which consists of the named arguments passed to `brew uses`. We're often only interested in the case where this is an `Array` of only one `Formula`, but `brew uses` also works for multiple named arguments, so we need to handle that case.

Reply:
So here's what I tried:
```diff
diff --git a/Library/Homebrew/dependencies_helpers.rb b/Library/Homebrew/dependencies_helpers.rb
index f967ab5b8..9c9dc3c4a 100644
--- a/Library/Homebrew/dependencies_helpers.rb
+++ b/Library/Homebrew/dependencies_helpers.rb
@@ -49,7 +49,7 @@ module DependenciesHelpers
     cache_key << "_#{root_dependent}_no_recursive_build_#{used_formulae.join("_")}" if skip_recursive_build_dependents
     cache_key.freeze
 
-    klass.expand(root_dependent, cache_key: cache_key) do |dependent, dep|
+    klass.expand(root_dependent, cache_key: nil) do |dependent, dep|
       if dep.recommended?
         klass.prune if ignores.include?("recommended?") || dependent.build.without?(dep)
       elsif dep.optional?
@@ -62,9 +62,9 @@ module DependenciesHelpers
 
         next unless skip_recursive_build_dependents
         next unless dep.build?
-        next if dependent == root_dependent && used_formulae.include?(dep.to_formula)
+        next if dependent != root_dependent
 
-        klass.prune
+        klass.keep_but_prune_recursive_deps
       end
 
       # If a tap isn't installed, we can't find the dependencies of one of
```
I killed caching to make sure I didn't just get the caching part wrong. Then:
```
❯ formula="go"; comm -23 <(brew uses --include-build --include-test --recursive --formula "$formula") <(brew uses --include-build --include-test --recursive --formula --skip-recursive-build-dependents "$formula") | tee godepsdiffstesting
arduino-cli
bat-extras
crun
envoy
envoy@1.18
kubernetes-cli
modd
newrelic-infra-agent
terraform@0.11
```

This is missing a bunch of formulae, like
```
❯ brew deps --tree --include-build --include-test --annotate kubectx
kubectx
└── kubernetes-cli
    ├── bash  [build]
    ├── coreutils  [build]
    │   └── gmp
    └── go@1.16  [build]
        └── go  [build]
```
Am I doing this wrong?

Reply:
Similar issues with this patch (which would have been nice since it would be better for caching):
```diff
diff --git a/Library/Homebrew/dependencies_helpers.rb b/Library/Homebrew/dependencies_helpers.rb
index f967ab5b8..1a74984c5 100644
--- a/Library/Homebrew/dependencies_helpers.rb
+++ b/Library/Homebrew/dependencies_helpers.rb
@@ -49,7 +49,7 @@ module DependenciesHelpers
     cache_key << "_#{root_dependent}_no_recursive_build_#{used_formulae.join("_")}" if skip_recursive_build_dependents
     cache_key.freeze
 
-    klass.expand(root_dependent, cache_key: cache_key) do |dependent, dep|
+    klass.expand(root_dependent, cache_key: nil) do |dependent, dep|
       if dep.recommended?
         klass.prune if ignores.include?("recommended?") || dependent.build.without?(dep)
       elsif dep.optional?
@@ -62,9 +62,8 @@ module DependenciesHelpers
 
         next unless skip_recursive_build_dependents
         next unless dep.build?
-        next if dependent == root_dependent && used_formulae.include?(dep.to_formula)
 
-        klass.prune
+        klass.keep_but_prune_recursive_deps
       end
 
       # If a tap isn't installed, we can't find the dependencies of one of
```

Reply:
But the basic idea of the change in this PR is that we prune any build dependent that isn't a direct dependent of a formula passed to `brew uses` (these are the formulae in `used_formulae`).

Reply:
Apologies - I'll be a little slow in responding until the weekend.

I think I might be thinking about this differently then you. I thought the intention was to avoid this:

```
C
└── B [build]
    └── A  [build]
```

But if I'm getting this correctly you actually want to drop that regardless if B is build or runtime. In which case, you can probably reduce the complexity by simply not passing `build?` to `includes` and appending a non-recursive `d.deps.select { |dep| dep.build? }` to the result. This won't require any changes to `dependencies_helpers` at all.

Reply:
Possibly. Might be as simple as `f.recursive_dependencies.reject(&:build?) + f.deps.selecrt(&:build?)`?